### PR TITLE
GH-40507: [C++][ORC] Upgrade ORC to 2.0.0

### PR DESCRIPTION
**Pull request by wgtmac**
### Rationale for this change

This PR aims to upgrade to a new major version of Apache ORC: https://orc.apache.org/news/2024/03/08/ORC-2.0.0/

### What changes are included in this PR?

This PR upgrades ORC dependency from 1.9.2 to 2.0.0.

### Are these changes tested?

Pass the CIs.

### Are there any user-facing changes?

No.
* GitHub Issue: #40507

### CODE REVIEW

**Comment by wgtmac** [comment]
OK, I'm wrong:

```
[INFO] Arrow Performance Benchmarks ....................... SKIPPED
[INFO] Arrow Java C Data Interface ........................ SUCCESS [ 57.354 s]
[INFO] Arrow Orc Adapter .................................. FAILURE [ 35.491 s]
[INFO] Arrow Gandiva ...................................... SUCCESS [01:26 min]
[INFO] Arrow Java Dataset ................................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  06:03 min (Wall Clock)
[INFO] Finished at: 2024-03-13T15:23:50Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.12.1:testCompile (default-testCompile) on project arrow-orc: Compilation failure
Error:  /arrow/java/adapter/orc/src/test/java/org/apache/arrow/adapter/orc/OrcReaderTest.java:[40,21] error: cannot access OrcFile
Error:    bad class file: /root/.m2/repository/org/apache/orc/orc-core/2.0.0/orc-core-2.0.0.jar(/org/apache/orc/OrcFile.class)
Error:      class file has wrong version 61.0, should be 53.0
Error:      Please remove or make sure it appears in the correct subdirectory of the classpath.
Error:  -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
Error:  
Error:  After correcting the problems, you can resume the build with the command
Error:    mvn <args> -rf :arrow-orc
```

**Comment by lidavidm** [comment]
Hmm, that is test scope, and the other builds pass. So I guess that pipeline uses Java 8 and we could maybe exclude the file?

**Comment by lidavidm** [comment]
That said, I think what we've done is to hold off on package updates requiring Java 8 for now, since we're still supporting Java 8...We were trying to drop Java 8 support but some people spoke out against. Maybe we need to revisit that.

**Comment by wgtmac** [comment]
As `orc-core` is only used in the test scope, perhaps we can leave it to stay at 1.9.x which is the last major version with java 8 support.

**Comment by lidavidm** [comment]
I added this to the list at https://github.com/apache/arrow/issues/38051

**Comment by lidavidm** [comment]
I think what's going to have to happen is that if someone really needs Java 8 support, they are going to have to maintain the fork or otherwise come up with a strategy, because eventually these CVE fixes pile up and we won't be able to defer the issue any longer

**Comment by wgtmac** [comment]
@github-actions crossbow submit wheel-windows-*

**Comment by lidavidm** [comment]
I think we're going to revisit Java 8 in the next 3-6 months so we can include this PR in that effort.

**Review by pitrou — CHANGES_REQUESTED** [review]
@wgtmac Can you please remove the Java change so that this PR focusses on the C++ version bump?

File: java/adapter/orc/pom.xml
Comment (wgtmac) [inline java/adapter/orc/pom.xml:41]:
```suggestion
            <version>1.9.2</version>
```

**Comment by wgtmac** [comment]
Just reverted the java module change. Thanks for the reminder! @pitrou

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python -g wheel

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
So it seems this breaks the macOS wheel builds for Python 3.8, 3.9 and 3.10 (but not 3.11 and 3.12).
The error is related to a protobuf symbol, and ORC uses protobuf:
```
ImportError: dlopen(/Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/test-arm64-env/lib/python3.8/site-packages/pyarrow/lib.cpython-38-darwin.so, 0x0002): symbol not found in flat namespace '__ZN6google8protobuf2io19ZeroCopyInputStream8ReadCordEPN4absl12lts_202401164CordEi'
```

The symbol is `google::protobuf::io::ZeroCopyInputStream::ReadCord(absl::lts_20240116::Cord*, int)`.

@wgtmac Is ORC 2.0 building its own protobuf from scratch?

**Comment by pitrou** [comment]
I'm trying to find any visible differences between https://github.com/ursacomputing/crossbow/actions/runs/8464875206/job/23191249944 (successful 3.10 build) and https://github.com/ursacomputing/crossbow/actions/runs/8464875206/job/23190409628 (failing 3.10 build) but nothing obvious stands out. The library versions seem exactly identical.

I'm going to restart a second time to see if we get a new failure.

**Comment by pitrou** [comment]
(I'll note that both ORC 1.9.3 and 2.0.0 seem to use `google::protobuf::io::ZeroCopyInputStream`, so it's not a new feature)

**Comment by pitrou** [comment]
So, I grabbed two 3.8 wheel files: one successful and one failed. Those wheels contain differences in the three following files: `libarrow.1600.dylib`, `libarrow_flight.1600.dylib` and `libarrow_substrait.1600.dylib`.

I'm going to try to find out more about this, but there doesn't seem to be an easy way to inspect macOS dylib files under Linux.

**Comment by pitrou** [comment]
@github-actions crossbow submit wheel-macos-*-arm64

**Comment by pitrou** [comment]
I also see the following CMake warning which is possibly related:
```
2024-03-28T09:13:53.3314130Z CMake Warning:
2024-03-28T09:13:53.3314590Z   Manually-specified variables were not used by the project:
2024-03-28T09:13:53.3314850Z 
2024-03-28T09:13:53.3314940Z     ORC_PROTOBUF_EXECUTABLE
2024-03-28T09:13:53.3315080Z 
```

**Comment by pitrou** [comment]
@github-actions crossbow submit wheel-macos-*-arm64

**Comment by pitrou** [comment]
Okay, I think I've found it:
* ORC build on [successful wheel build](https://productionresultssa15.blob.core.windows.net/actions-results/543076d3-56af-4fbc-8260-78e45bc3fda2/workflow-job-run-ca395085-040a-526b-2ce8-bdc85f692774/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-03-28T13%3A52%3A05Z&sig=XlDLT8YXosUno5bsHkMdQAJUgREfZstBGDDo0xWoaCA%3D&sp=r&spr=https&sr=b&st=2024-03-28T13%3A42%3A00Z&sv=2021-12-02):
```
2024-03-28T13:34:25.3459520Z -- PROTOBUF_HOME: /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/vcpkg/installed/arm64-osx-static-release
2024-03-28T13:34:25.3460120Z -- Could NOT find Protobuf (missing: Protobuf_DIR)
2024-03-28T13:34:25.3460830Z -- Found the Protobuf headers: /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/vcpkg/installed/arm64-osx-static-release/include
2024-03-28T13:34:25.3461910Z -- Found the Protobuf library: /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/vcpkg/installed/arm64-osx-static-release/lib/libprotobuf.a
2024-03-28T13:34:25.3463000Z -- Found the Protoc library: /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/vcpkg/installed/arm64-osx-static-release/lib/libprotoc.a
2024-03-28T13:34:25.3464060Z -- Found the Protoc executable: /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/vcpkg/installed/arm64-osx/tools/protobuf/protoc
2024-03-28T13:34:25.3465220Z -- Found the Protobuf static library: /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/vcpkg/installed/arm64-osx-static-release/lib/libprotobuf.a
2024-03-28T13:34:25.3466370Z -- Found the Protoc static library: /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/vcpkg/installed/arm64-osx-static-release/lib/libprotoc.a
```
* ORC build on [failed wheel build](https://productionresultssa9.blob.core.windows.net/actions-results/f4aa3d21-d458-443e-a829-9102cd02db40/workflow-job-run-ca395085-040a-526b-2ce8-bdc85f692774/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-03-28T13%3A46%3A19Z&sig=QUAj6fWpvA%2FnyLNK8Duevklu8hJ1GkpwXfH67Th4wwY%3D&sp=r&spr=https&sr=b&st=2024-03-28T13%3A36%3A14Z&sv=2021-12-02):
```
2024-03-28T13:33:59.5638710Z -- PROTOBUF_HOME: /Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/vcpkg/installed/arm64-osx-static-release
2024-03-28T13:33:59.5643750Z -- Found the Protobuf headers: /opt/homebrew/include
2024-03-28T13:33:59.5644210Z -- Found the Protobuf library: protobuf::libprotobuf
2024-03-28T13:33:59.5644590Z -- Found the Protoc library: protobuf::libprotoc
2024-03-28T13:33:59.5644960Z -- Found the Protoc executable: protobuf::protoc
```

Note `/opt/homebrew/include`.

**Comment by pitrou** [comment]
@github-actions crossbow submit wheel-macos-*-arm64

**Comment by pitrou** [comment]
Yuck, this PR is based on an old version of the git repo...

**Comment by pitrou** [comment]
@github-actions crossbow submit wheel-macos-*-arm64

**Comment by wgtmac** [comment]
> So it seems this breaks the macOS wheel builds for Python 3.8, 3.9 and 3.10 (but not 3.11 and 3.12). The error is related to a protobuf symbol, and ORC uses protobuf:
> 
> ```
> ImportError: dlopen(/Users/voltrondata/github-actions-runner/_work/crossbow/crossbow/test-arm64-env/lib/python3.8/site-packages/pyarrow/lib.cpython-38-darwin.so, 0x0002): symbol not found in flat namespace '__ZN6google8protobuf2io19ZeroCopyInputStream8ReadCordEPN4absl12lts_202401164CordEi'
> ```
> 
> The symbol is `google::protobuf::io::ZeroCopyInputStream::ReadCord(absl::lts_20240116::Cord*, int)`.
> 
> @wgtmac Is ORC 2.0 building its own protobuf from scratch? Answer: it should not, we're passing `PROTOBUF_HOME` to ORC's CMake.

Right, ORC should use what we provide here: https://github.com/apache/arrow/blob/a9b2cc2c962f064c3fa5504909f122e9bcabda3f/cpp/cmake_modules/ThirdpartyToolchain.cmake#L4386-L4396

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python -g wheel -g cpp

**Review by pitrou — APPROVED** [review]
CI results look unrelated now, I'll merge

**Comment by wgtmac** [comment]
Thank you for the investigation! @pitrou

**Pull request merged**
